### PR TITLE
Update Vehicle Locations Use Case to Accept dirTag Parameter

### LINKED ISSUE #58: Update Vehicle Locations use case with Directions
# Update Vehicle Locations Use Case to Accept `dirTag` Parameter

## Description
The current implementation of the Vehicle Locations use case lacks the flexibility to filter vehicles based on specific direction parameters. To enhance the functionality and provide more targeted information, the Vehicle Locations use case must be updated to accept an additional `dirTag` as a parameter, in addition to the existing `routeTag` parameter. This change will allow users to retrieve a list of vehicles that are specifically on a particular route and direction.

## Proposed Changes
- Modify the Vehicle Locations use case to include `routeTag` and `dirTag` parameters in its method signature.
- Update the implementation to filter vehicles based on the provided route and direction tags.
- Ensure that the use case returns a List of Vehicles that match the specified route and direction.

### PR DESCRIPTION
**Pull request by nsaroiu**
# Pull Request: Update Vehicle Locations Use Case to Accept `dirTag` Parameter

## Issue
Closes #58

## Description
This pull request addresses the issue of enhancing the Vehicle Locations use case by allowing it to accept an additional `dirTag` parameter for more targeted vehicle information.

## Changes Made
- Updated `VehicleLocationsController` to take in an additional `dirTag` parameter.
- Updated `VehicleLocationsService` to accept the new `dirTag` parameter.
- Modified `VehicleLocationsImplementation` to filter vehicles based on both `routeTag` and `dirTag`.
- Added comprehensive tests for the Vehicle Locations use case, covering various scenarios:
  - Case 1: Returns a nonempty list of Vehicles.
  - Case 2: Returns an empty list of Vehicles when no vehicles exist with the given `dirTag`.
  - Case 3: Returns an empty list of Vehicles when no vehicles exist for a given `routeTag`.
  - Case 4: Returns an empty list of Vehicles when the `routeTag` does not exist.

## Testing
- All new tests pass successfully.
- Existing tests remain unaffected.

## Reviewer Notes
Please pay attention to the modifications made in the `VehicleLocationsController`, `VehicleLocationsService`, and `VehicleLocationsImplementation` files. Additionally, review the new test cases to ensure they cover the mentioned scenarios effectively.

## Checklist
- [x] Code follows the project's coding standards.
- [x] All tests pass.
- [x] Documentation is updated.
- [x] The commit history is clean and follows best practices.

### CODE REVIEW

File: backend/src/main/java/com/example/backend/interface_adapters/VehicleLocationsController.java
Comment (prash-red) [inline backend/src/main/java/com/example/backend/interface_adapters/VehicleLocationsController.java:1]: You have added the dirTag as a parameter on this endpoint. This is a good design choice.

File: backend/src/main/java/com/example/backend/use_case/vehicle_locations/VehicleLocationsImplementation.java
Comment (prash-red) [inline backend/src/main/java/com/example/backend/use_case/vehicle_locations/VehicleLocationsImplementation.java:1]: I understand that you get all the vehicles for a particular route and filter through the ones that go through the given direction. This takes O(n) time. Is there a way to improve the efficiency of this implementation ?

File: backend/src/main/java/com/example/backend/use_case/vehicle_locations/VehicleLocationsService.java
Comment (prash-red) [inline backend/src/main/java/com/example/backend/use_case/vehicle_locations/VehicleLocationsService.java:1]: This is synonymous with the Implementation.

File: backend/src/test/java/com/example/backend/use_case/vehicle_locations/VehicleLocationsImplementationTest.java
Comment (prash-red) [inline backend/src/test/java/com/example/backend/use_case/vehicle_locations/VehicleLocationsImplementationTest.java:1]: This is too long to read

File: backend/src/main/java/com/example/backend/use_case/vehicle_locations/VehicleLocationsImplementation.java
Comment (nsaroiu) [reply backend/src/main/java/com/example/backend/use_case/vehicle_locations/VehicleLocationsImplementation.java:1]: Dude its O(n) for a max n of like 50.

File: backend/src/main/java/com/example/backend/use_case/vehicle_locations/VehicleLocationsService.java
Comment (nsaroiu) [reply backend/src/main/java/com/example/backend/use_case/vehicle_locations/VehicleLocationsService.java:1]: Bro it's called abstraction

**Review by prash-red — APPROVED** [review]
This code adheres to the CA engine and the documentation is very good. Next time make sure that the curly braces are consistent.

**Pull request merged**